eZ/Publish/Core/Persistence/Legacy/Content/Gateway/DoctrineDatabase.php

```diff
@@ -1891,4 +1893,167 @@ public function copyRelations($originalContentId, $copiedContentId, $versionNo =
+                    ':languageCode' => $languageCode,
+                    ':contentId' => $contentId,
+                ]
+            )
+        ;
+
+        $query->execute();
+    }
+
+    /**
+     * Remove language from language_mask of ezcontentobject.
+     *
+     * @param int $contentId
+     * @param int $languageId
+     * @throws \eZ\Publish\Core\Base\Exceptions\BadStateException
+     */
+    private function deleteTranslationFromContentObject($contentId, $languageId)
+    {
+        $query = $this->connection->createQueryBuilder();
+        $query->update('ezcontentobject')
+            // parameter for bitwise operation has to be placed verbatim (w/o binding) for this to work cross-DBMS
+            ->set('language_mask', 'language_mask & ~ ' . $languageId)
+            ->set('modified', ':now')
+            ->where('id = :contentId')
+            ->andWhere(
+            // make sure removed translation is not the last one (incl. alwaysAvailable)
+                $query->expr()->orX(
+                    'language_mask & ~ ' . $languageId . ' <> 0',
+                    'language_mask & ~ ' . $languageId . ' <> 1'
+                )
+            )
+            ->setParameter(':now', time())
+            ->setParameter(':contentId', $contentId)
+        ;
+
+        $rowCount = $query->execute();
+
+        // no rows updated means that most likely somehow it was the last remaining translation
+        if ($rowCount === 0) {
+            throw new BadStateException(
+                '$languageCode',
+                'Specified translation is the only one Content Object Version has'
+            );
+        }
+    }
+
+    /**
+     * Remove language from language_mask of ezcontentobject_version and update initialLanguageId
+     * if it matches the removed one.
+     *
+     * @param int $contentId
+     * @param int $languageId
+     * @throws \eZ\Publish\Core\Base\Exceptions\BadStateException
+     */
+    private function deleteTranslationFromContentVersions($contentId, $languageId)
+    {
+        $mainLanguageSelectQuery = $this->connection->createQueryBuilder();
+        $mainLanguageSelectQuery
+            ->select('initial_language_id')
+            ->from('ezcontentobject', 'c')
+            // parameter is to be bound in the main query
+            ->where('c.id = :contentId')
+        ;
+
+        $query = $this->connection->createQueryBuilder();
+        $query->update('ezcontentobject_version')
+            // parameter for bitwise operation has to be placed verbatim (w/o binding) for this to work cross-DBMS
+            ->set('language_mask', 'language_mask & ~ ' . $languageId)
+            ->set('modified', ':now')
+            // update initial_language_id only if it matches removed translation languageId
+            ->set(
+                'initial_language_id',
+                'CASE WHEN initial_language_id = :languageId THEN (' . $mainLanguageSelectQuery->getSQL() . ') ELSE initial_language_id END'
```

Comment:
> Would it be possible to do like in copyRelations() a few methods above here? Imo that approach would be cleaner when we need to do things like this :)

Ok, I've discussed this w/ @Nattfarinn and we've agreed that this should be done in the same way. So, with better indentation for readability, amended in 2a3a400.